test/wallet-rescan-test.js

```diff
@@ -0,0 +1,516 @@
+/* eslint-env mocha */
+/* eslint prefer-arrow-callback: "off" */
+/* eslint no-implicit-coercion: "off" */
+
+'use strict';
+
+const assert = require('bsert');
+const FullNode = require('../lib/node/fullnode');
+const MemWallet = require('./util/memwallet');
+const Network = require('../lib/protocol/network');
+const Address = require('../lib/primitives/address');
+const rules = require('../lib/covenants/rules');
+const {Resource} = require('../lib/dns/resource');
+const {forValue} = require('./util/common');
+
+const network = Network.get('regtest');
+
+const {
+  treeInterval,
+  biddingPeriod,
+  revealPeriod,
+  transferLockup
+} = network.names;
+
+describe('Wallet rescan with namestate transitions', function() {
+  describe('Only sends OPEN', function() {
+    // Bob runs a full node with wallet plugin
+    const node = new FullNode({
+      network: network.type,
+      memory: true,
+      plugins: [require('../lib/wallet/plugin')]
+    });
+    node.on('error', (err) => {
+      assert(false, err);
+    });
+
+    const {wdb} = node.require('walletdb');
+    let bob, bobAddr;
+
+    // Alice is some other wallet on the network
+    const alice = new MemWallet({ network });
+    const aliceAddr = alice.getAddress();
+
+    // Connect MemWallet to chain as minimally as possible
+    node.chain.on('connect', (entry, block) => {
+      alice.addBlock(entry, block.txs);
+    });
+    alice.getNameStatus = async (nameHash) => {
```

Comment:
this might be a nitpick but why is this function being defined on an alice object when it's a pure function? is it an override?

Reply:
yes, `alice` here is a test utility called `MemWallet` which is just a dumb transaction generator. Its not the actual wallet being tested (`Bob` is -- with a fullnode) and so `MemWallet` is missing some functions and doesn't have access to chain data so when it needs to place an open for example, we need to stub this function.

Reply:
ok good to know, makes sense